pkg/query/endpointset.go

```diff
@@ -590,7 +595,8 @@ type endpointRef struct {
 // newEndpointRef creates a new endpointRef with a gRPC channel to the given the IP address.
 // The call to newEndpointRef will return an error if establishing the channel fails.
 func (e *EndpointSet) newEndpointRef(ctx context.Context, spec *GRPCEndpointSpec) (*endpointRef, error) {
-	conn, err := grpc.DialContext(ctx, spec.Addr(), e.dialOpts...)
+	dialOpts := append(e.dialOpts, spec.dialOpts...)
+	conn, err := grpc.DialContext(ctx, spec.Addr(), dialOpts...)
```

Comment:
Afaiu, this would mean that any group of Thanos endpoints can be configured to be load-balanced. But we don't check if those components are actually HA via same labels or so. 

So in case of faulty configuration, debugging might be confusing here.

Reply:
That is correct, and I think this is not something we can easily solve. It seems to be one of the drawbacks of using the built-in load balancing in gRPC alongside the max connection age workaround.